File: docs/JAMIA/comparisons/RPython/TM/22906804_matrix_multiply_dataframe.R
# ID: SO#22906804
# Language: R
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: Matrix multiply on data.frame requires as.matrix. Type not suitable for math operations.
# Reproduction status: Live
# Type system catch: `Matrix` type required for `%*%`; `DataFrame` rejected

da <- data.frame(
  V1 = c(0.46, 0.25, 0.82),
  V2 = c(2.36, 1.52, 1.50),
  V3 = c(-1.54, -0.59, 0.34)
)

# BUG: %*% requires matrix, not data.frame
tryCatch(
  t(da) %*% da,
  error = function(e) cat("Crash:", conditionMessage(e), "\n")
)
# Error: requires numeric/complex matrix/vector arguments

# Fix: convert to matrix first
ma <- diag(3) + t(as.matrix(da)) %*% as.matrix(da)
print(ma)

File: docs/JAMIA/comparisons/RPython/TM/22906804_matrix_multiply_dataframe.ts
/**
 * RPython SO#22906804 — Matrix expression causes "requires numeric/complex matrix/vector arguments"
 * Effect: Crash
 * Bug class: Type coercion
 *
 * In R, %*% on a data.frame crashes — requires as.matrix() first. Nothing enforces
 * this at compile time.
 *
 * In tidy-ts, DataFrame columns are typed arrays. You extract number[] columns
 * explicitly for numeric operations. There is no ambiguity between DataFrame and
 * matrix — they are different things and the type system separates them.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const da = createDataFrame([
  { v1: 0.46, v2: 2.36, v3: -1.54 },
  { v1: 0.25, v2: 1.52, v3: -0.59 },
  { v1: 0.82, v2: 1.50, v3: 0.34 },
]);

// @ts-expect-error — DataFrame is not a number[] — can't pass it where array expected
const wrong = s.mean(da);

// Extract typed arrays for numeric operations
const v1 = da.extract("v1"); // number[]
const v2 = da.extract("v2"); // number[]
console.log(s.mean(v1), s.mean(v2));
